src/common/dav/tx.c

```diff
@@ -699,21 +735,28 @@ dav_tx_end(void)
 
 	if (tx->stage_callback &&
 			(tx->stage == DAV_TX_STAGE_ONCOMMIT ||
-			tx->stage == DAV_TX_STAGE_ONABORT)) {
+			 tx->stage == DAV_TX_STAGE_ONABORT)) {
 		tx->stage = DAV_TX_STAGE_FINALLY;
 		obj_tx_callback(tx);
 	}
 
-	struct tx_data *txd = PMDK_SLIST_FIRST(&tx->tx_entries);
+	struct tx_data *txd = DAV_SLIST_FIRST(&tx->tx_entries);
 
-	PMDK_SLIST_REMOVE_HEAD(&tx->tx_entries, tx_entry);
+	DAV_SLIST_REMOVE_HEAD(&tx->tx_entries, tx_entry);
 
 	D_FREE(txd);
 
 	VALGRIND_END_TX;
 	int ret = tx->last_errnum;
 
-	if (PMDK_SLIST_EMPTY(&tx->tx_entries)) {
+	if (DAV_SLIST_EMPTY(&tx->tx_entries)) {
+		DAV_DBG("");
+		ASSERT(tx->pop);
+
+		/* commit to WAL */
+		if (ret == 0)
+			dav_wal_tx_commit(tx->pop);
```

Comment:
wal_commit() will yield, it should be performed after setting tx->stage as NONE, also, the TLS for storing current TX should be cleared before wal_commit(). Please see my comment about per transaction TX, we need to allow multiple TXs run in parallel. (You could refer to ad_tx.c code to see how we allocate TX and set/unset TLS)

Reply:
wal_commit() will yield, it should be performed after setting tx->stage as NONE, absolutely. Accepted.